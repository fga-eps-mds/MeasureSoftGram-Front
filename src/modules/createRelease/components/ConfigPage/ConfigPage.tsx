import React, { useEffect } from 'react';

import { Box, Checkbox, Switch, Typography } from '@mui/material';


import { useCreateReleaseContext } from '@modules/createRelease/context/useCreateRelease';
import ConfigsForm from '../ConfigsForm';
import CONFIG_PAGE from './consts';
import * as Styles from './styles';

const { SUB_TITLE } = CONFIG_PAGE;

interface ConfigPageProps {
  page: number;
  title: string;
}

const ConfigPage = ({
  page,
  title,
}: ConfigPageProps) => {
  const {
    setCurrentConfig,
    configPageData,
    changeThreshold,
    toggleChangeThreshold,
    allowChangeConfig,
    toggleAllowChangeConfig
  } = useCreateReleaseContext();
  const {
    characteristicCheckbox,
    setCharacteristicCheckbox,
    characteristicData,
    setSubcharacterCheckbox,
    subcharacterCheckbox,
    setMeasureCheckbox,
    measureCheckbox,
    setCharacteristicValuesValid
  } = configPageData;

  useEffect(() => {
    setCharacteristicValuesValid(false);
  }, [page]);

  const renderPage = () => {
    const genericProps = { onChange: setCurrentConfig, setIsValuesValid: setCharacteristicValuesValid, data: characteristicData };
    const subtitle = 'Definir os pesos das';
    if (page === 0) {
      return (
        <ConfigsForm
          {...genericProps}
          subtitle={`${subtitle} características`}
          type="characteristic"
          disable={!allowChangeConfig}
          checkboxValues={characteristicCheckbox}
          setCheckboxValues={setCharacteristicCheckbox}
        />
      );
    }
    if (page === 1) {
      return (
        <ConfigsForm
          {...genericProps}
          tabs={characteristicCheckbox}
          type="subcharacteristic"
          disable={!allowChangeConfig}
          subtitle={`${subtitle} subcaracterísticas`}
          checkboxValues={subcharacterCheckbox}
          setCheckboxValues={setSubcharacterCheckbox}
        />
      );
    }
    return (
      <Box display="flex" flexDirection="column">
        <ConfigsForm
          {...genericProps}
          tabs={subcharacterCheckbox}
          subtitle={`${subtitle} medidas`}
          type="measure"
          disable={!allowChangeConfig}
          checkboxValues={measureCheckbox}
          setCheckboxValues={setMeasureCheckbox}
        />
        <Typography color="#FF4646" marginTop="40px">
          <Checkbox
            checked={changeThreshold}
            onChange={toggleChangeThreshold}
            disabled={!allowChangeConfig}
            name="changeThresholdCheckbox"
            color="secondary"
          />
          Modificar valores de referência mínimo e máximo
        </Typography>
      </Box>
    );
  };

  return (
    <>
      <Styles.Header>
        <h1 style={{ color: '#33568E', fontWeight: '500' }}>{title}</h1>
      </Styles.Header>
      <Styles.Body>
        <p>{SUB_TITLE}</p>
        <p>
          <Switch checked={allowChangeConfig} onChange={toggleAllowChangeConfig} color="secondary" />
          Editar configurações
        </p>
      </Styles.Body>
      {renderPage()}
    </>
  );
};

export default ConfigPage;
