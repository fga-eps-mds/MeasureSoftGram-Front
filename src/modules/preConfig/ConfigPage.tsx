import React, { useEffect, useState } from 'react';

import { Alert, Box, Checkbox, Snackbar, Switch, Typography } from '@mui/material';

import { productQuery } from '@services/product';

import { Characteristic } from '@customTypes/preConfig';

import { useCreateReleaseContext } from '@modules/createRelease/context/useCreateRelease';
import ConfigsForm from './components/ConfigsForm';
import { useQuery } from './hooks/useQuery';
import CONFIG_PAGE from './consts';
import * as Styles from './styles';

const { TITLE, SUB_TITLE, DEFAULT_MESSAGE, ERROR_MESSAGE } = CONFIG_PAGE;

interface ConfigPageProps {
  isOpen: boolean;
  onClose: Function;
  repoName?: string;
  organizationId: string;
  productId: string;
  page: number;
  filteredCharacteristics: string[];
  title: string;
}

const ConfigPage = ({
  repoName = '',
  organizationId,
  productId,
  page,
  filteredCharacteristics,
  title
}: ConfigPageProps) => {
  const request = useQuery();
  const {
    setCurrentConfig,
    configPageData,
    changeThreshold,
    toggleChangeThreshold,
    allowChangeConfig,
    toggleAllowChangeConfig
  } = useCreateReleaseContext();
  const {
    characteristicCheckbox,
    setCharacteristicCheckbox,
    characteristicData,
    setSubcharacterCheckbox,
    subcharacterCheckbox,
    setMeasureCheckbox,
    measureCheckbox
  } = configPageData;

  const [data, setData] = useState(
    characteristicData.length
      ? characteristicData
      : request?.data.characteristics.filter((c: Characteristic) => filteredCharacteristics.includes(c.key))
  );

  const [isValuesValid, setIsValuesValid] = useState(false);
  const [error, setError] = useState(false);
  const [showAlert, setShowAlert] = useState(false);

  useEffect(() => {
    setIsValuesValid(false);
  }, [page]);

  useEffect(() => {
    setCurrentConfig(data);
  }, [data]);

  useEffect(() => { }, []);

  const clearRequest = (jsonValue: Characteristic[]): Characteristic[] => {
    const jsonString = JSON.stringify(jsonValue);
    const jsonStringReplaced = jsonString?.replace(/\d+/g, '0');
    return JSON?.parse(jsonStringReplaced) as Characteristic[];
  };

  const isArrayNull = (array: Array<any>) => array.length === 0;

  const isFormCompleted = () => {
    if (isArrayNull(characteristicCheckbox) && page === 0) return true;
    if (isArrayNull(subcharacterCheckbox) && page === 1) return true;
    if (isArrayNull(measureCheckbox) && page === 2) return true;
  };

  const sendJson = () => {
    const responseCharacterFiltered = data?.filter((charcterValue) =>
      characteristicCheckbox.includes(charcterValue.key)
    );
    const responseSubcharacterFiltered = responseCharacterFiltered?.map((charcterValue) => ({
      ...charcterValue,
      subcharacteristics: charcterValue.subcharacteristics.filter((subcharcterValue) =>
        subcharacterCheckbox.includes(subcharcterValue.key)
      )
    }));
    const finalData = responseSubcharacterFiltered?.map((charcterValue) => ({
      ...charcterValue,
      subcharacteristics: charcterValue.subcharacteristics.map((subcharcterValue) => ({
        ...subcharcterValue,
        measures: subcharcterValue.measures.filter((measureValue) => measureCheckbox.includes(measureValue.key))
      }))
    })) as Characteristic[];

    setShowAlert(true);
    productQuery
      .postPreConfig(organizationId, productId, { name: repoName, data: { characteristics: finalData } })
      .catch(() => {
        setError(true);
      });
  };

  const renderPage = () => {
    const genericProps = { onChange: setData, setIsValuesValid, data };
    const subtitle = 'Definir os pesos das';
    if (page === 0) {
      return (
        <ConfigsForm
          {...genericProps}
          subtitle={`${subtitle} caracteristicas`}
          type="characteristic"
          disable={!allowChangeConfig}
          checkboxValues={characteristicCheckbox}
          setCheckboxValues={setCharacteristicCheckbox}
        />
      );
    }
    if (page === 1) {
      return (
        <ConfigsForm
          {...genericProps}
          tabs={characteristicCheckbox}
          type="subcharacteristic"
          disable={!allowChangeConfig}
          subtitle={`${subtitle} subcaracteristicas`}
          checkboxValues={subcharacterCheckbox}
          setCheckboxValues={setSubcharacterCheckbox}
        />
      );
    }
    return (
      <Box display="flex" flexDirection="column">
        <ConfigsForm
          {...genericProps}
          tabs={subcharacterCheckbox}
          subtitle={`${subtitle} medidas`}
          type="measure"
          disable={!allowChangeConfig}
          checkboxValues={measureCheckbox}
          setCheckboxValues={setMeasureCheckbox}
        />
        <Typography color="#FF4646" marginTop="40px">
          <Checkbox
            checked={changeThreshold}
            onChange={toggleChangeThreshold}
            disabled={!allowChangeConfig}
            name="changeThresholdCheckbox"
          />
          Modificar valores de referência mínimo e máximo
        </Typography>
      </Box>
    );
  };

  return (
    <>
      <Styles.Header>
        <h1>{title}</h1>
      </Styles.Header>
      <Styles.Body>
        <p>{SUB_TITLE}</p>
        <p>
          <Switch checked={allowChangeConfig} onChange={toggleAllowChangeConfig} />
          Editar configurações
        </p>
      </Styles.Body>
      {renderPage()}
    </>
  );
};

export default ConfigPage;
