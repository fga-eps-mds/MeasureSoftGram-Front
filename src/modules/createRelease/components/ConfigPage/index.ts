import ConfigPage from './ConfigPage';

export default ConfigPage;
