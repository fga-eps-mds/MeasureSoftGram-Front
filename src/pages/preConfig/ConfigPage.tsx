import DrawerMenu from '@components/DrawerMenu';
import { Characteristic } from '@customTypes/preConfig';
import { ButtonType } from '@customTypes/project';
import { Alert, Snackbar, Typography } from '@mui/material';
import { projectQuery } from '@services/project';
import React, { useEffect, useState } from 'react';
import ConfigsForm from './components/ConfigsForm';
import { useQuery } from './hooks/useQuery';
import CONFIG_PAGE from './consts';

const { TITLE, SUB_TITLE, DEFAULT_MESSAGE, ERROR_MESSAGE } = CONFIG_PAGE;

interface ConfigPageProps {
  isOpen: boolean;
  onClose: Function;
  repoName: string;
}

const ConfigPage = ({ isOpen, onClose, repoName }: ConfigPageProps) => {
  const request = useQuery();

  const [data, setData] = useState(request?.data.characteristics);
  const [characterCheckbox, setCharacterCheckbox] = useState<string[]>([]);
  const [subcharacterCheckbox, setSubcharacterCheckbox] = useState<string[]>([]);
  const [measureCheckbox, setMeasureheckbox] = useState<string[]>([]);
  const [page, setPage] = useState(0);
  const [isValuesValid, setIsValuesValid] = useState(false);
  const [error, setError] = useState(false);
  const [showAlert, setShowAlert] = useState(false);

  useEffect(() => {
    setPage(0);
    setData(clearRequest(data));

    setCharacterCheckbox([]);
    setSubcharacterCheckbox([]);
    setMeasureheckbox([]);
  }, [isOpen]);

  useEffect(() => {
    setIsValuesValid(false);
  }, [page]);

  useEffect(() => {}, []);

  const clearRequest = (jsonValue: Characteristic[]): Characteristic[] => {
    const jsonString = JSON.stringify(jsonValue);
    const jsonStringReplaced = jsonString?.replace(/\d+/g, '0');
    return JSON?.parse(jsonStringReplaced) as Characteristic[];
  };

  const isArrayNull = (array: Array<any>) => array.length === 0;

  const isFormCompleted = () => {
    if (isArrayNull(characterCheckbox) && page === 0) return true;
    if (isArrayNull(subcharacterCheckbox) && page === 1) return true;
    if (isArrayNull(measureCheckbox) && page === 2) return true;
  };

  const sendJson = () => {
    const responseCharacterFiltered = data?.filter((charcterValue) => characterCheckbox.includes(charcterValue.key));
    const responseSubcharacterFiltered = responseCharacterFiltered?.map((charcterValue) => ({
      ...charcterValue,
      subcharacteristics: charcterValue.subcharacteristics.filter((subcharcterValue) =>
        subcharacterCheckbox.includes(subcharcterValue.key)
      )
    }));
    const finalData = responseSubcharacterFiltered?.map((charcterValue) => ({
      ...charcterValue,
      subcharacteristics: charcterValue.subcharacteristics.map((subcharcterValue) => ({
        ...subcharcterValue,
        measures: subcharcterValue.measures.filter((measureValue) => measureCheckbox.includes(measureValue.key))
      }))
    })) as Characteristic[];

    setShowAlert(true);
    projectQuery.postPreConfig('1', '1', { name: repoName, data: { characteristics: finalData } }).catch(() => {
      setError(true);
    });
  };

  const renderNextOrEndButton = (): ButtonType => {
    let label = 'AvanÃ§ar';
    let onClick = () => {
      const nextPage = page <= 2 ? page + 1 : 0;
      setPage(nextPage);
    };
    if (page === 2) {
      label = 'Finalizar';
      onClick = () => {
        onClose(false);
        sendJson();
      };
    }
    return {
      label,
      onClick,
      disabled: isFormCompleted() || !isValuesValid,
      backgroundColor: 'white',
      color: 'black',
      variant: 'outlined'
    };
  };

  const buttons: Array<ButtonType> = [
    {
      label: 'Cancelar',
      onClick: () => {
        setData(clearRequest(data));
        onClose(false);
      },
      backgroundColor: 'black',
      color: 'white'
    },
    {
      label: 'Retornar',
      onClick: () => {
        const previousPage = page >= 0 ? page - 1 : 2;
        setPage(previousPage);
      },
      backgroundColor: 'gray',
      disabled: page === 0,
      color: 'white'
    },
    renderNextOrEndButton()
  ];

  const renderPage = () => {
    const genericProps = { onChange: setData, setIsValuesValid, data };
    const subtitle = 'Definir os pesos das';
    if (page === 0) {
      return (
        <ConfigsForm
          {...genericProps}
          subtitle={`${subtitle} caracteristicas`}
          type="characteristic"
          checkboxValues={characterCheckbox}
          setCheckboxValues={setCharacterCheckbox}
        />
      );
    }
    if (page === 1) {
      return (
        <ConfigsForm
          {...genericProps}
          tabs={characterCheckbox}
          type="subcharacteristic"
          subtitle={`${subtitle} subcaracteristicas`}
          checkboxValues={subcharacterCheckbox}
          setCheckboxValues={setSubcharacterCheckbox}
        />
      );
    }
    return (
      <ConfigsForm
        {...genericProps}
        tabs={subcharacterCheckbox}
        subtitle={`${subtitle} medidas`}
        type="measure"
        checkboxValues={measureCheckbox}
        setCheckboxValues={setMeasureheckbox}
      />
    );
  };

  return (
    <>
      <DrawerMenu open={isOpen} buttons={buttons} title={TITLE} subtitle={SUB_TITLE}>
        <>
          <Typography variant="h6" mt="24px">
            {repoName}
          </Typography>
          {renderPage()}
        </>
      </DrawerMenu>
      <Snackbar
        open={showAlert}
        autoHideDuration={6000}
        onClose={() => {
          setShowAlert(false);
          setError(false);
        }}
      >
        <Alert severity={error ? 'error' : 'success'} sx={{ width: '100%' }}>
          {error ? ERROR_MESSAGE : DEFAULT_MESSAGE}
        </Alert>
      </Snackbar>
    </>
  );
};

export default ConfigPage;
