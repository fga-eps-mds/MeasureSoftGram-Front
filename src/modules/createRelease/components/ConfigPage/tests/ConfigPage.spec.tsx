import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';

import ConfigPage from '../ConfigPage';

const REPO_NAME = 'REPO_TITLE';

jest.mock('@services/product', () => ({
  productQuery: {
    getProductCurrentPreConfig: () => ({ data: { "id": 25, "name": "3", "data": { "characteristics": [{ "key": "reliability", "weight": 50, "subcharacteristics": [{ "key": "testing_status", "weight": 100, "measures": [{ "key": "passed_tests", "weight": 33 }, { "key": "test_builds", "weight": 33 }, { "key": "test_coverage", "weight": 34 }] }] }, { "key": "maintainability", "weight": 50, "subcharacteristics": [{ "key": "modifiability", "weight": 100, "measures": [{ "key": "non_complex_file_density", "weight": 33 }, { "key": "commented_file_density", "weight": 33 }, { "key": "duplication_absense", "weight": 34 }] }] }] }, "created_at": "2022-09-15T18:56:57-03:00" } }),
    getPreConfigEntitiesRelationship: () => ({
      data: [{
        key: 'reliability',
        id: 1,
        name: 'Confiabilidade',
        description: '',
        subcharacteristics: []
      }]
    }),
    createProductReleaseGoal: () => ({ data: [{ id: 1, release_name: 'r1', start_at: '', end_at: '', changes: [] }] })
  }
}));

jest.mock('@modules/createRelease/context/useCreateRelease', () => ({
  ...jest.requireActual('@modules/createRelease/context/useCreateRelease'),
  useCreateReleaseContext: () => ({
    configPageData: {
      characteristicCheckbox: [],
      setCharacteristicCheckbox: () => { },
      characteristicData: [],
      setCharacteristicValuesValid: (value: boolean) => { },
    },
    setCharacteristicValuesValid: (value: boolean) => { },
  })
}));

describe('<ConfigPage />', () => {
  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <ConfigPage page={0} title={REPO_NAME} />
      );
      expect(tree).toMatchSnapshot();
    });
  });
});
