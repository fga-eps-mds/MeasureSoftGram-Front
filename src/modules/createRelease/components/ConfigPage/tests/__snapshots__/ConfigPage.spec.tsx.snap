// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfigPage /> Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="styles__Header-sc-vaumvh-0 bzwHPl"
      >
        <h1>
          REPO_TITLE
        </h1>
      </div>
      <div
        class="styles__Body-sc-vaumvh-1 cBkRoe"
      >
        <p>
          Por meio da pré-configuração que é definido quais características, subcaracterísticas e medidas serão consideradas para mensurar a qualidade do software analisado. Na pré-configuração também é definido os pesos de cada uma das entidades definidas.
        </p>
        <p>
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
          Editar configurações
        </p>
      </div>
      <div
        class="MuiBox-root css-2c3mdi"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1mo68eo-MuiTypography-root"
        >
          Definir os pesos das caracteristicas
        </h6>
        <div
          class="MuiGrid-root MuiGrid-container css-s3g6yz-MuiGrid-root"
        />
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="styles__Header-sc-vaumvh-0 bzwHPl"
    >
      <h1>
        REPO_TITLE
      </h1>
    </div>
    <div
      class="styles__Body-sc-vaumvh-1 cBkRoe"
    >
      <p>
        Por meio da pré-configuração que é definido quais características, subcaracterísticas e medidas serão consideradas para mensurar a qualidade do software analisado. Na pré-configuração também é definido os pesos de cada uma das entidades definidas.
      </p>
      <p>
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
          />
        </span>
        Editar configurações
      </p>
    </div>
    <div
      class="MuiBox-root css-2c3mdi"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 css-1mo68eo-MuiTypography-root"
      >
        Definir os pesos das caracteristicas
      </h6>
      <div
        class="MuiGrid-root MuiGrid-container css-s3g6yz-MuiGrid-root"
      />
      <div
        class="MuiBox-root css-0"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
